### PR TITLE
Google TTS Cache Management Fix

### LINKED ISSUE #3819: GoogleTTS cache handling
## Expected Behavior
Calls to voice synthesis should use the new configuration after the configuration has been changed instead of returning old results from the cache.

## Current Behavior
Cache need to be manually purged to resynthesize old sentences with the new configuration.

## Possible Solution
* Cached file names could encode the configuration details as well
* Cache could be purged after the configuration was changed.

### PR DESCRIPTION
**Pull request by gbicskei**
Cache management fix
* Cache can be purged from Configuration/System/Voice configuration screen
* Cache file names contain the voice name instead of the locale
* Technical parameters are encoded into the cache file name

Fixes #3819

Signed-off-by: Gabor Bicskei <gbicskei@gmail.com>

### CODE REVIEW

File: addons/voice/org.openhab.voice.googletts/ESH-INF/config/config.xml [hunk 1]
```diff
@@ -27,6 +27,11 @@
 			<description>Speaking rate can be 4x faster or slower than the normal rate</description>
 			<default>1</default>
 		</parameter>
+		<parameter name="purgeCache" type="boolean">
+			<label>Purge the cache</label>
+			<description>Purge the cache after configuration changes</description>
```
Comment (wborn) [inline addons/voice/org.openhab.voice.googletts/ESH-INF/config/config.xml]: I think it should be emphasized here that after enabling it, users also need to disable it again. Otherwise users change it to true, continue with their life and after restarting openHAB their caches are always purged. 

It would have been nice if users could just press a button to purge the cache. But since ESH only allows for changing configuration values this is not possible.

Another option would be to change the setting back to false from within the binding after having purged the caches. E.g. using the OSGi `ConfigurationAdmin` service. Though I do not see any other extension change their own settings with it. So that could introduce an architectural issue.

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java [hunk 2]
```diff
@@ -229,14 +230,37 @@ File synthesizeSpeech(String text, GoogleTTSVoice voice, String codec) {
...
+        FileOutputStream fos = new FileOutputStream(cacheFile);
+        fos.write(audio);
+        fos.close();
+        logger.debug("Caching audio file {}", cacheFile.getName());
+
+        // write text to file for transparency too
+        // this allows to know which contents is in which audio file
+        String txtFileName = FilenameUtils.removeExtension(cacheFile.getName()) + ".txt";
+        FileOutputStream txtFos = new FileOutputStream(new File(cacheFolder, txtFileName));
+        txtFos.write(text.getBytes(UTF_8));
```
Comment (wborn) [inline addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: AFAIK these text files are used as meta data for the cached synthesized audio data. This information could then be used to regenerate or purge (specific) content. I think that won't be possible without also storing the parameters that are used in `GoogleTTSConfig.toConfigString`?

The VoiceRSS TTS addon (where I think this caching implementation originates from) doesn't have this issue. It has  no additional settings and has only one voice per locale.

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java [hunk 3]
```diff
@@ -229,14 +230,37 @@ File synthesizeSpeech(String text, GoogleTTSVoice voice, String codec) {
         }
     }
 
+    /**
+     * Create cache entry.
+     *
+     * @param text      Converted text.
+     * @param cacheFile Cache entry file.
+     * @param audio     Byte array of the audio.
+     * @throws IOException in case of file handling exceptions
```
Comment (wborn) [inline addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: The formatter in Eclipse will put only one space between the parameter names and description in JavaDocs. So can you adjust your IDE settings?

So whenever some saves this file in Eclipse this is automatically formatted to:

```java
    /**
     * Create cache entry.
     *
     * @param text Converted text.
     * @param cacheFile Cache entry file.
     * @param audio Byte array of the audio.
     * @throws IOException in case of file handling exceptions
     */
```

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java [hunk 4]
```diff
@@ -105,13 +102,23 @@ void setConfig(GoogleTTSConfig config) {
                 voices.clear();
             }
         }
+
+        //maintain cache
+        if (config.getPurgeCache() != null && config.getPurgeCache()) {
+            File[] files = cacheFolder.listFiles();
+            if (files != null && files.length > 0) {
+                //noinspection ResultOfMethodCallIgnored
```
Comment (wborn) [inline addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: Looks like these are some kind of IntelliJ specific annotations similar to `@SuppressWarnings`. I don't see them used anywhere else in the openHAB/ESH code base. Can you change your IDE settings to suppress the warnings and remove these comments?

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java (hunk 2 above)
Comment (gbicskei) [reply addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: Regenerating everything would create a peak load here which I don't want. Texts will be regenerated  anyway when they are referenced... Also not changing the code for this. Ok?

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java (hunk 3 above)
Comment (gbicskei) [reply addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: Sure. I'll reformat the entire binding...

**Comment by gbicskei** [comment]
@wborn 
Is there anything else I can do with this PR?

**Comment by martinvw** [comment]
@wborn given you already did a review, could you validate whether its fine now?

File: addons/voice/org.openhab.voice.googletts/ESH-INF/config/config.xml (hunk 1 above)
Comment (wborn) [reply addons/voice/org.openhab.voice.googletts/ESH-INF/config/config.xml]: I'd rather also have this added to the configuration parameter description because in my own experience people can't find or don't read the documentation. :wink: 

After looking at the implementation I'd rephrase the description to:

> Purges the cache e.g. after testing different voice configuration parameters. When enabled the cache is purged once. Make sure to disable this setting again so the cache is maintained after restarts.

Comment (gbicskei) [reply addons/voice/org.openhab.voice.googletts/ESH-INF/config/config.xml]: Description added

File: addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java (hunk 2 above)
Comment (gbicskei) [reply addons/voice/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java]: I added the config parameters and the voice name to the txt file.

File: addons/voice/org.openhab.voice.googletts/README.md [hunk 5]
```diff
@@ -46,6 +46,7 @@ Using your favourite configuration UI (e.g. PaperUI) edit **Services/Voice/Googl
 * **Pitch** - The pitch of selected voice, up to 20 semitones
 * **Volume Gain** - The volume of the output between 16dB and -96dB
 * **Speaking Rate** - The speaking rate can be 4x faster or slower than the normal rate 
+* **Purge Cache** - Purge the cache content after the configuration is saved. This flag is useful while playing with different voices. **When the configuration stable make sure to set this flag to false to enable the binding to use the cache.**
```
Comment (wborn) [inline addons/voice/org.openhab.voice.googletts/README.md]: Can you add the same description here as used in config.xml?

> Purges the cache e.g. after testing different voice configuration parameters. When enabled the cache is purged once. Make sure to disable this setting again so the cache is maintained after restarts.

Comment (gbicskei) [reply addons/voice/org.openhab.voice.googletts/README.md]: Done

**Comment by cweitkamp** [comment]
Does it make sense to reset the `purgeCache` configuration automatically after the clean has been performed?

**Comment by wborn** [comment]
It would be nice but I think the addon would need to depend on the  OSGi `ConfigurationAdmin` service for that. I could not find any other addon that does this. Probably because addons should not update configurations.

It would be even better to just have a button in the UI that calls a method to purge the cache.
